### PR TITLE
Track ongoing tasks in TaskExecutionTimeTrackingEsThreadPoolExecutor

### PR DESCRIPTION
**Pull request by pxsalehi**
For Stateless autoscaling, to be able to account for ongoing tasks in
the estimated ingestion load, we need to keep track of the ongoing
tasks. This is important because currently the only stats that we get
from the executor is the `TotalTaskExecutionTime` which is updated only
once a task is finished. This can lead to cases where if we do have long
tasks (write tasks) that last longer than a sampling period (1s), we'd
observe no write load and assume there is no load, because the tasks
would be included in the stats only once they are finished. Recording
ongoing tasks allows taking these into consideration.

Relates ES-6391

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:133]: Note that for the estimation we do not need the tasks counted/considered in `ongoingTasks` and `totalExecutionTime` to be exclusive. This also helps to avoid unnecessary synchronization.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:113]: I think it's better to wrap the whole body of the function in a `try...finally` and put this line in the finally part. Because I see that `super.afterExecute(r, t);` may throw, and possibly who knows other stuff may throw, so it is better to ensure that the removal happens with the finally part.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:89]: Do you need the accuracy of `System.nanoTime()` or would you be fine with the more performant but less accurate thread cached time (`absoluteTimeInMillis()`)?

**Review by kingherc — COMMENTED** [review]
Some comments.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java]: It does sound safer. Pushed [bb63339](https://github.com/elastic/elasticsearch/pull/97661/commits/bb63339ff9b0d160cfc4d14d312bc9d9bf31fdff).

File: server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:89]: The wrapper (`TimedRunnable`) that records the total execution time of the task is using `System.nanoTime()`, which is why I've also chosen the same time source to avoid any surprises. I mean we could reuse the recorded value there by unwrapping and exposing the recorded start time, not sure if that just complicates things, but it's an option.

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:89]: Actually let me try...

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java]: I've pushed https://github.com/elastic/elasticsearch/pull/97661/commits/35529f837562736c94252e91a1e936bf5180a80f.

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:90]: I've reverted https://github.com/elastic/elasticsearch/commit/35529f837562736c94252e91a1e936bf5180a80f and added a comment in https://github.com/elastic/elasticsearch/pull/97661/commits/d37ba5d950574ed1409ed5446cfcf985b77ed305. I think reusing the timestamp of the wrapper is not correct. That timestamp starts when the task in submitted to the executor. We want the timestamp of when it actually starts execution which could be (much) later, therefore we need to capture the time again. And I'd rather use the same source.

Comment (kingherc) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java]: Ah I had not realized that TimedRunnable is already using System.nanoTime(). Then LGTM

**Review by kingherc — APPROVED** [review]
LGTM after handling my couple of comments. Feel free to think about getting an approval from one of the colleagues with the more experience in the subject.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:146]: Updating the ongoing tasks start time doesn't seem right to me, since it just records the start time. At the end, in `AverageWriteLoadSampler` we do cap the time that we estimate for each task and for all ongoing tasks per sampling period.

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/TaskExecutionTimeTrackingEsThreadPoolExecutor.java:146]: As discussed, the double counting scenario is a known side-effect of not synchronizing the total execution time and ongoing tasks, but for the estimation we're ok with it, due to the capping. If we find this at some point not accurate enough, we could address it by incurring some extra overhead , e.g. using the approach suggested by Francisco.

**Review by fcofdez — APPROVED** [review]
LGTM

**Pull request merged**
